### PR TITLE
Revisit back pressure implementation

### LINKED ISSUE #35703: Make Pipe limits not throw when hitting backpressure and instead call Complete on the writer awaitable
Today, when hitting backpressure when calling AdvanceReader in Pipes, we normally would just throw saying we hit backpressure preventing a deadlock. This has a few issues though:

- Whoever is calling AdvanceTo on the pipe has no idea whether the call will hit backpressure or not. 
- There is simply no way to use a Pipe to buffer something larger than the pause/reader threshold.

The change would involve removing the call to throw here and instead call `_writerAwaitable.Complete(out completionData);`
https://github.com/dotnet/corefx/blob/master/src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L492-L495

This would change the behavior from throwing to thrashing, which isn't much better behavior than before. 

cc @halter73 @Tratcher @davidfowl @pakrym

### PR DESCRIPTION
**Pull request by davidfowl**
- Today when using a PipeReader from a Pipe, when the pause threshold is hit and the reader does not consume enough data to unblock the writer (enough to reduce the resume threshold) an exception is thrown. This exception is extremely hard to understand and work around since it's based on a limit set by the owner of the Pipe. It leads to people setting an extremely big limit (bigger than the biggest message) or disabling the limit completely. It's especially problematic when library authors need to expose a PipeReader but don't know what upstack protocol will be parsed and in turn doesn't know the max message size.
- The changes the back pressure mechanic to be based on examined instead of consumed. This pauses the writer if the pause threshold is hit, but lets the reader continue to buffer until a logical payoad is reached (whatever matters to the protocol).
- **This breaks software that relied on these limits to enfore a maximum buffer size and those pieces of code need to now handle their own maximum buffer size.**

Fixes #35703

cc @mgravell @benaadams @tmds

### CODE REVIEW

File: src/System.IO.Pipelines/tests/PipeLengthTests.cs [hunk 1]
```diff
@@ -87,13 +87,13 @@ public async Task LengthDecreasedAfterReadAdvanceConsume()
...
         [Fact]
-        public async Task LengthNotChangeAfterReadAdvanceExamine()
+        public async Task LengthDoesNotChangeIfExamineDoesNotChange()
         {
             PipeWriter writableBuffer = _pipe.Writer.WriteEmpty(10);
             await writableBuffer.FlushAsync();
 
             ReadResult result = await _pipe.Reader.ReadAsync();
-            _pipe.Reader.AdvanceTo(result.Buffer.Start, result.Buffer.End);
+            _pipe.Reader.AdvanceTo(result.Buffer.Start, result.Buffer.Start);
```
Comment (jkotalik) [inline src/System.IO.Pipelines/tests/PipeLengthTests.cs]: Just make this a single Buffer.Start

File: src/System.IO.Pipelines/tests/BackpressureTests.cs [hunk 2]
```diff
@@ -29,39 +29,6 @@ public void Dispose()
 
         private readonly Pipe _pipe;
 
-        [Fact]
-        public void AdvanceThrowsIfFlushActiveAndNotConsumedPastThreshold()
```
Comment (pakrym) [inline src/System.IO.Pipelines/tests/BackpressureTests.cs]: Turn this into tests for the new behavior. We don't have any.

File: src/System.IO.Pipelines/tests/PipeLengthTests.cs [hunk 3]
```diff
@@ -87,13 +87,13 @@ public async Task LengthDecreasedAfterReadAdvanceConsume()
         }
 
         [Fact]
-        public async Task LengthNotChangeAfterReadAdvanceExamine()
```
Comment (pakrym) [inline src/System.IO.Pipelines/tests/PipeLengthTests.cs]: Add ExaminingWithoutConsumingChangesLenght test.

File: src/System.IO.Pipelines/tests/BackpressureTests.cs (hunk 2 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/tests/BackpressureTests.cs]: We kinda do though, the ones that test backpressure today pass the same thing for consumed an examined.

Comment (pakrym) [reply src/System.IO.Pipelines/tests/BackpressureTests.cs]: No. We need a test that consumes nothing but advances everything past the backpressure limit and make sure writer is unblocked.

Comment (pakrym) [reply src/System.IO.Pipelines/tests/BackpressureTests.cs]: We only have tests that either consume and examine at the same time or not hit backpressure.

Comment (davidfowl) [reply src/System.IO.Pipelines/tests/BackpressureTests.cs]: Sure, that sounds like a single test 😄 .

Comment (davidfowl) [reply src/System.IO.Pipelines/tests/BackpressureTests.cs]: Wrote a more complicated test and found a bug

**Comment by davidfowl** [comment]
🆙 📅

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 4]
```diff
@@ -60,15 +60,20 @@ public sealed partial class Pipe
         private PipeCompletion _writerCompletion;
         private PipeCompletion _readerCompletion;
 
-        // The read head which is the extent of the IPipelineReader's consumed bytes
+        // Stores the last examined position, used to calculate how much bytes were to release
```
Comment (jkotalik) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]:
```suggestion
        // Stores the last examined position, used to calculate how many bytes to release
```

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 5]
```diff
@@ -60,15 +60,20 @@ public sealed partial class Pipe
...
+        // for back pressure management
+        private BufferSegment _lastExamined;
```
Comment (jkotalik) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: How much do we care about increasing the size of the pipe with these fields?

File: src/System.IO.Pipelines/tests/BackpressureTests.cs [hunk 6]
```diff
@@ -30,55 +30,80 @@ public void Dispose()
...
+
+                    Assert.False(flushTask.IsCompleted);
+
+                    await flushTask;
+                }
+
+                _pipe.Writer.Complete();
+            }
+
+           Task writingTask = WriteLoopAsync();
```
Comment (jkotalik) [inline src/System.IO.Pipelines/tests/BackpressureTests.cs]: nit space

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs [hunk 7]
```diff
@@ -488,11 +511,6 @@ private void AdvanceReader(BufferSegment consumedSegment, int consumedIndex, Buf
                 // but only if writer is not completed yet
                 if (examinedEverything && !_writerCompletion.IsCompleted)
                 {
-                    // Prevent deadlock where reader awaits new data and writer await backpressure
-                    if (!_writerAwaitable.IsCompleted)
```
Comment (jkotalik) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Can we Debug.Assert that _writerAwaitable.IsCompleted?

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs (hunk 5 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: I care but I haven't found a more succinct way to implement this yet.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs (hunk 7 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs]: Sure

File: src/System.IO.Pipelines/tests/BackpressureTests.cs [hunk 8]
```diff
@@ -84,7 +84,7 @@ async Task WriteLoopAsync()
                 _pipe.Writer.Complete();
             }
 
-           Task writingTask = WriteLoopAsync();
+            Task writingTask = WriteLoopAsync();
```
Comment (pakrym) [inline src/System.IO.Pipelines/tests/BackpressureTests.cs]: Nooooooooo

**Comment by pakrym** [comment]
:shipit:

File: src/System.IO.Pipelines/tests/BackpressureTests.cs (hunk 8 above)
Comment (jkotalik) [reply src/System.IO.Pipelines/tests/BackpressureTests.cs]: ??????????

Comment (pakrym) [reply src/System.IO.Pipelines/tests/BackpressureTests.cs]: I thought this was an extra space.

**Comment by tmds** [comment]
@davidfowl perhaps this now also enables the pipe to be configured with minimal values (i.e. ` 1`). Then the pipe won't buffer more data than what is needed by the application, which reduces the overall memory used by the application. That is what I was trying when I [first hit this issue](https://github.com/dotnet/corefxlab/issues/1355#issuecomment-288955908).

**Pull request merged**